### ISSUE TITLE
3D QG

### ISSUE THREAD

**Issue by dhruvbalwada** [pyqg/pyqg#246]
Was just wondering how easy/difficult would it be to translate @shfer 's 3D QG model (https://cims.nyu.edu/~shafer/tools/index.html) to be part of the pyqg suite?

**Comment by rabernat** [comment]
I certainly think it is possible. How easy/difficult depends on a number of details. The big question is whether the model can be implemented as a subclass of the existing base Model class, or whether it would require modifications to the kernel.

**Comment by shfer** [comment]
I think pyqg already contains most of the essential functionality of my model.  What aspect of it would you like to see imported @dhruvbalwada ?

**Comment by dhruvbalwada** [comment]
I was just mulling over your 2009 paper and wondering about 3D structures that form, and thinking that it would be nice if some results were reproducible. However, I assume it is essentially the multilayer version that is part of pyqg (?).

**Comment by rabernat** [comment]
Multi-layer and multi-level are a bit different. We currently have implemented multi-layer. Multi-level is presumably very similar. Pyqg just requires you to define an inversion relationship between the streamfunction, which advects the PV at each level / layer, and the PV itself.  For the layered model, this is where that happens:

https://github.com/pyqg/pyqg/blob/6bc7416edb201a52db61940999855d19feebb41b/pyqg/layered_model.py#L205-L223

As long as you can frame the system in these terms, it's relatively easy to support in pyqg. One type of model that _would not_  be easily supported is a modal model, because the advection relationship between modes is more complicated. Pyqg requires that advection be "local"; the PV at each "level" (whatever level happens to mean for that model) is advected by the streamfunction at that level.

**Comment by shfer** [comment]
I spent too much time thinking about this at some point.  My final take is
that it turns out that there is very little difference between the layer
and level versions. In the latter, one should have explicit representation
of the conservation of buoyancy at the upper and lower surfaces, and at the
level of finite differencing, the upper surface buoyancy is bT = f0 (ψ1-ψ2
)/Δ1.  In the layer version, the stretching term in the top layer is almost
identical to this, which means the advection of the upper layer PV is
almost advection of q1+bT.  At the end of the day I find an extra factor of
2 shows up, and the difference in what you get by extracting bT from the
layer version, vs. the explicit bT in the level version is negligible.

AND, anyway, if you're interested in SF09, that was really about interior
dynamics so these details above are irrelevant.

Long story short: I don't think any changes to pyqg are necessary to
continue similar experiments (but you do need a couple tracers)

**Comment by francispoulin** [comment]
To get a 3D QG model that ignores buoyancy at the surface, you need to have a 3D inversion operator.  The advection is the same and happens at each level, as it happens at each layer in the layered version.

Has anyone tried doing a 3D inversion in pyqg?  I'm happy to try it but thought I would see if anyone has already tried this or not.
